### ISSUE TITLE
Official ubuntu package

### ISSUE THREAD

**Issue by benbro** [openjudge/sandbox#6]
Is there a chance libsandbox packages will be included in the official ubuntu repository?
It's important that a security package will have more testers and credibility.

Thanks

**Comment by liuyu81** [comment]
We would love to see more participants join and help us evaluate / improve `libsandbox`. But we have to point out that `libsandbox` is _not_ a general-purpose security package.

Strictly speaking, `libsandbox` is a configurable profiling / instrumenting tool for educational system developers. Its code base originates from an automated grading system for educational programming contests; and up till now, the majority of users are still from educational background, including developers of course management systems, organizers of programming contests, and instructors of undergraduate-level IT/CS courses. 

To best address the need of these targeted users, `libsandbox` was designed as a highly-configurable, user-space API library. And such design choice comes with several limitations (which are listed in the README file). First, `libsandbox` only supports single-process programs (because multi-process programs can break user-space system call auditing through intentional race conditions); Second, `libsandbox` does not ship with pre-defined policies (rather, we allow our users to customize policies that are _secure_ for their individual requirements); Third, `libsandbox` has been carefully tuned for profiling accuracy and consistency on virtualized platforms (such as Amazon EC2, VirtualBox, and Xen nodes).

With these design choices in mind, we believe that not everyone needs to install `libsandbox` on their own computers -- albeit some of them may have used online educational services powered by `libsandbox`. Proposing `libsandbox` to the official repository of a mainstream Linux distribution is both unnecessary and infeasible.

If there is considerable demands, we may set up separate, public repositories to host the latest deb and rpm packages of `libsandbox`. When such a repository is available, we will post a notice on our web site at http://openjudge.net/.

**Comment by benbro** [comment]
Will I be able to use libsandbox with programs like postgresql, redis, octave or are all of them multi-process programs?
How does libsandbox compares to lxc-execute or libvirt-sandbox?

If you submit libsandbox to the official ubuntu repository as openjudge-sandbox it will be clear that it's not a general purpose sandbox. What If I'll install the deb package on my server to execute untrusted student code but the deb package itself will introduce malicious code or bug? When a package is in the official repository it is tested and reviewed by a large number of developers and users.

The problem I have with libsandbox right now is that although it's an open source project, a large part of it is hidden. The mailing list is private, there are no docs, it's not clear where is the official git repo, commit messages are cryptic and the deb package is pre-compiled.

Thanks

**Comment by liuyu81** [comment]
The philosophy of `libsandbox` is that we do _not_ define security for our users. `libsandbox` just captures activities of the target program as events and dispatches them to user-specified policies to decide which are acceptable and which are malicious. As a small open source project, we cannot afford to develop and maintain policy rules for all those third-party applications in the wild -- therefore we provide API's for our users to define policy rules that they feel appropriate. For example, if opening file `abc` for write is considered malicious, users of `libsandbox` can forbid this activity with a policy rule. We have no idea if postgres, redis, or octave are multi-process programs or not. But perhaps you can `strace` them to find out by yourself.

Comparing `libsandbox` with OS-level virtualization tools such as `libvirt-sandbox` is like comparing `SELinux` with `Java`'s policy-based security mechanism -- they employ different approaches of sandboxing and work at different levels of granularities (see Daniel P. Berrangé's talk at FOSDEM 2012). `libsandbox` is a user-space API library for profiling and instrumenting simple programs, and it implements Linux system call interception. `libvirt-sandbox` is an OS-level isolation tool that creates virtual containers for the targeted program, in which all the hardware/software resources available to the targeted program are isolated from the hosting OS. As a matter of fact, `libsandbox` was never designed to compete with virtualization-based sandboxing tools -- it can run _within_ virtualized environments.

`libsandbox` is distributed in terms of the New BSD license. To use it, you are at your own risk. We cannot guarantee that `libsandbox` is bug-free, or appropriate for your application, or what-so-over. But it is open for everyone to try, evaluate, and modify. The `deb` packages available at http://sourceforge.net/projects/libsandbox are always derived from the source tar balls (which are in sync with _this_ repository). In fact, `deb` packages are converted from `rpm` packages using `alien`, and `rpm` packages are built from source tar balls with `rpmbuild`. 

You are right in that the development repository of `libsandbox` is not public. This is because we are serious about releasing new versions, and we do not want work-in-progress to propagate. New releases are available to the public only after internal stress tests. Docs are missing? The document of `libsandbox` is embedded within the source code and can be generated with `doxygen`. If you are a Python developer, you can always type `help(sandbox)` in python's interactive console to view the document of `pysandbox`. Commit log's are cryptic? We do not use commit logs to record changes, we log changes in files named `CHANGES` (available in the source tar balls). Mailing list is private? No, it is open for everyone to join -- just apply for membership and give us a hint on why you are interested.

**Comment by benbro** [comment]
It could be nice to have the mailing list available at least in read-only mode.
Maybe you could have a public list and a security list where only approved members can view.
Please consider some kind of public interface because this is a great project but I feel that it prevents community interaction.

Can you please share your build script or makefile that generate the rpm packages with rpmbuild and the deb packages with alien so we could build it ourselves?

Thanks

**Comment by liuyu81** [comment]
for `libsandbox` the core library

`$ rpmbuild -tb libsandbox-<version>.tar.gz`

for `pysandbox` the pythonic wrapper

`$ tar -xzvf pysandbox-<version>.tar.gz && cd pysandbox-<version>`
`$ python setup.py bdist_rpm`

when you have rpm packages at hand

`$ sudo alien -k *.rpm`

to create `libsandbox` tar ball from git checkout

`$ cd libsandbox`
`$ ./configure --enable-chkvsc`
`$ make sdist`

to create `pysandbox` tar ball from git checkout

`$ cd pysandbox`
`$ python setup.py sdist`

**Comment by benbro** [comment]
This is extremely helpful.
Thank you.
